manticore/core/smtlib/solver.py

```diff
@@ -432,20 +395,95 @@ def can_be_true(self, constraints: ConstraintSet, expression: Union[bool, Bool]
+        :param max_iter: maximum number of iterations allowed
+        """
+        # TODO: consider adding a mode to return best known value on timeout
+        assert goal in ("maximize", "minimize")
+        operation = {"maximize": Operators.UGE, "minimize": Operators.ULE}[goal]
+
+        last_value: Optional[Union[int, bool, bytes]] = None
+
+        start = time.time()
+        with constraints as temp_cs:
+            X = temp_cs.new_bitvec(x.size)
+            temp_cs.add(X == x)
+            aux = temp_cs.new_bitvec(X.size, name="optimized_")
+            self._reset(temp_cs.to_string(related_to=X))
+
+            # Find one value and use it as currently known min/Max
+            if not self._is_sat():
+                raise SolverException("UNSAT")
+            last_value = self._getvalue(X)
+            self._assert(operation(X, last_value))
+
+            # This uses a binary search to find a suitable range for aux
+            # Use known solution as min or max depending on the goal
+            if goal == "maximize":
+                m, M = last_value, (1 << x.size) - 1
+            else:
+                m, M = 0, last_value
+
+            # Iteratively divide the range
+            L = None
+            while L not in (M, m):
+                L = (m + M) // 2
+                self._assert(operation(X, L))
+                sat = self._is_sat()
+
+                # depending on the goal move one of the extremes
+                if goal == "maximize" and sat or goal == "minimize" and not sat:
+                    m = L
+                else:
+                    M = L
+
+                if time.time() - start > consts.timeout:
+                    raise SolverError("Timeout")
+
+            # reset to before the dichotomic search
+            self._reset(temp_cs.to_string(related_to=X))
+
+            # At this point we know aux is inside [m,M]
+            # Lets constrain it to that range
+            self._assert(Operators.UGE(X, m))
+            self._assert(Operators.ULE(X, M))
+
+            # And now check all remaining possible extremes
+            last_value = None
+            i = 0
+            while self._is_sat():
+                last_value = self._getvalue(X)
+                self._assert(operation(X, last_value))
+                self._assert(X != last_value)
+                i = i + 1
+                if i > max_iter:
+                    raise SolverError("Optimizing error, maximum number of iterations was reached")
+                if time.time() - start > consts.timeout:
+                    raise SolverError("Timeout")
+            if last_value is not None:
+                return last_value
+            raise SolverError("Optimizing error, unsat or unknown core")
+
     @lru_cache(maxsize=32)
-    def get_all_values(self, constraints, expression, maxcnt=None, silent=False):
+    def get_all_values(
+        self, constraints: ConstraintSet, expression, maxcnt: bool = None, silent: bool = False
```

Comment:
`maxcnt` should be `Optional[bool]`. You'll have to import `Optional` at the top though: `from typing import Optional`
```diff
-        self, constraints: ConstraintSet, expression, maxcnt: bool = None, silent: bool = False
+        self, constraints: ConstraintSet, expression, maxcnt: Optional[bool] = None, silent: bool = False
```